### PR TITLE
SAMZA-1385: Coordination utils factory with distributed lock

### PR DESCRIPTION
**Pull request by sborya**
this PR includes some changes from another PR. I will re-merge it again, after the other PR is in.

### CODE REVIEW

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java [hunk 1]
```diff
@@ -233,32 +214,41 @@ public void waitForFinish() {
...
-          }
-        } finally {
-          coordinationUtils.reset();
-        }
-      } else {
-        // each application process will try creating the streams, which
-        // requires stream creation to be idempotent
+    if (intStreams.isEmpty()) {
+      LOG.info("Set of intermediate streams is empty. Nothing to create.");
+      System.out.println("1");
```
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: interesting output :). There are a bunch of System.out below which need to be removed.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java [hunk 2]
```diff
@@ -233,32 +214,41 @@ public void waitForFinish() {
...
+      return;
+    }
+
+    // Move the scope of coordination utils within stream creation to address long idle connection problem.
+    // Refer SAMZA-1385 for more details
+    String coordinationId = new ApplicationConfig(config).getGlobalAppId() + APPLICATION_RUNNER_ZK_PATH_SUFFIX;
```
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: Can we drop the name ZK in the constant APPLICATION_RUNNER_ZK_PATH_SUFFIX  in this patch? it's not zk specific now.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java [hunk 3]
```diff
@@ -233,32 +214,41 @@ public void waitForFinish() {
...
+    CoordinationUtils coordinationUtils =
+        CoordinationUtilsFactory.getCoordinationUtilsFactory(config).getCoordinationUtils(coordinationId, uid, config);
+    System.out.println(" create streams." + uid);
```
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: remove

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java [hunk 4]
```diff
@@ -233,32 +214,41 @@ public void waitForFinish() {
...
+    if (coordinationUtils == null) {
+      // each application process will try creating the streams, which
+      // requires stream creation to be idempotent
+      getStreamManager().createStreams(intStreams);
+      System.out.println("failed to create utils" + uid);
```
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: remove

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java [hunk 5]
```diff
@@ -233,32 +214,41 @@ public void waitForFinish() {
...
+      return;
+    }
+
+    DistributedLockWithState lockWithState = coordinationUtils.getLockWithState(planId);
+    System.out.println("Init NEEDED for streams" + uid);
```
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: remove

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java [hunk 6]
```diff
@@ -233,32 +214,41 @@ public void waitForFinish() {
...
+    try {
+      // check if the processor needs to go through leader election and stream creation
+      if (lockWithState.lockIfNotSet(1000, TimeUnit.MILLISECONDS)) {
+        System.out.println("CREATING STREAMS" + uid);
```
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: probably need some meaningful logging here, ilke log.info("lock {} is acquired", lockId)

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java [hunk 7]
```diff
@@ -233,32 +214,41 @@ public void waitForFinish() {
...
         getStreamManager().createStreams(intStreams);
+        lockWithState.unlockAndSet();
+      } else {
+        System.out.println("SOMEONE ELSE CREATING STREAMS" + uid);
```
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: meaningful logging

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java [hunk 8]
```diff
@@ -233,32 +214,41 @@ public void waitForFinish() {
...
       }
+    } catch (TimeoutException e) {
+      LOG.error("Failed to get the lock for stream initialization");
```
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: probably need to wrap e inside a SamzaException and rethrow here

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java (hunk 1 above)
Comment (sborya) [reply samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: system.outs will be removed.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java (hunk 2 above)
Comment (sborya) [reply samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: done.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java (hunk 3 above)
Comment (sborya) [reply samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: will do at merge.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java (hunk 4 above)
Comment (sborya) [reply samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: will do at merge.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java (hunk 5 above)
Comment (sborya) [reply samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: will do at merge.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java (hunk 6 above)
Comment (sborya) [reply samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: Done.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java (hunk 7 above)
Comment (sborya) [reply samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: will be done.

**Pull request closed without merging**